kong/db/dao/plugins.lua

```diff
@@ -309,4 +309,21 @@ function Plugins:load_plugin_schemas(plugin_set)
 end
 
 
+function Plugins:select_by_cache_key(key)
+  -- try new format
+  local entity, new_err = self.super.select_by_cache_key(self, key)
+  if entity then
+    return entity
+  end
+
+  -- try old format
+  local row, old_err = self.strategy:select_by_cache_key_migrating(key)
+  if row then
+    return self:row_to_entity(row)
+  end
+
+  return nil, (new_err and old_err) and new_err or old_err
```

Comment:
If we are trying the old format, isn't is most likely because `new_err` is set (e.g. to "invalid query: column 'cache_key' does not exist"). As such, wouldn't this condition supersede `old_err` (which may be a legitimate error from the legacy call) with this error from the new format call?

In short, shouldn't this be:
```lua
return nil, (new_err and old_err) and old_err or new_err
```

I am really unsure, maybe I missed another case this order covers.

Reply:
I scratched my head for a while on this too, and I don't think there's a perfect answer. Considering the cases:

1. row is found
   a. migrated: new format returns row
   b. unmigrated: old format returns row
2. there is no row
  a. migrated: new_err and old_err are nil
  b. unmigrated: new_err is "cache_key doesn't exist", old_err is nil
3. new format produces real error
  a. migrated: most likely old will error too? so both new_err and old_err will be set
  b. unmigrated: new_err is "cache_key doesn't exist", old_err is something (case you brought up)
4. old format produces real error
  a. migrated: new_err is nil, old_err has error
  b. unmigrated: new_err is "cache_key doesn't exist", old_err has the error

The original order gives a bad message on case 3b indeed. But flipping it as you suggest would fix 3b but produce a bad message on case 3a, _I think_. Since "migrated" is the common case, I went with that one.

When we manage to disable this translator after the migration is complete, then it would essentially run on the unmigrated case, so flipping the condition as you suggested would be ideal.